### PR TITLE
Chapter14

### PR DESCRIPTION
**Pull request by AndreeaSburlea**
## New features
- detail view

## Changes in code
- used custom fonts

## Screenshot
ch14

### CODE REVIEW

File: FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift [hunk 1]
```diff
@@ -9,20 +9,26 @@ import UIKit
 
 class RestaurantDetailViewController: UIViewController {
 
-    @IBOutlet private var restaurantImageView: UIImageView!
-    @IBOutlet private var restaurantName: UILabel!
-    @IBOutlet private var restaurantType: UILabel!
-    @IBOutlet private var restaurantLocation: UILabel!
+    @IBOutlet var tableView: UITableView!
+    @IBOutlet var headerView: RestaurantDetailHeaderView!
```
Comment (ralumarusca) [inline FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift]: outlets should be private

File: FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            }
+        }
+    }
+    @IBOutlet var typeLabel: UILabel! {
+        didSet {
+            if let customFont = UIFont(name: "Nunito-Bold", size: 20.0) {
+                typeLabel.font = UIFontMetrics(forTextStyle: .headline).scaledFont(for: customFont)
+            }
+        }
+    }
```
Comment (ralumarusca) [inline FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift]: IBOutlets should be private and the numberOfLines and font can also be set directly from storyboard

File: FoodPin/FoodPin/View/RestaurantDetailTextCell.swift [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+import UIKit
+
+class RestaurantDetailTextCell: UITableViewCell {
+
+    @IBOutlet var descriptionLabel: UILabel! {
+        didSet {
+            descriptionLabel.numberOfLines = 0
+        }
+    }
```
Comment (ralumarusca) [inline FoodPin/FoodPin/View/RestaurantDetailTextCell.swift]: same as previous comment

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            column2TitleLabel.text = column2TitleLabel.text?.uppercased()
+            column2TitleLabel.numberOfLines = 0
+        }
+    }
+
+    @IBOutlet var column2TextLabel: UILabel! {
+        didSet {
+            column2TextLabel.numberOfLines = 0
+        }
+    }
```
Comment (ralumarusca) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift]: Same as the previous comments and the text can be made uppercased directly where you initially set it

**Review by ralumarusca — CHANGES_REQUESTED** [review]

File: FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift (hunk 2 above)
Comment (AndreeaSburlea) [reply FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift]: I know about the numberOfLines and the font but in the book they teach us that can be done in both ways. Should I delete the lines of code and do it in the other way or to leave it like this?

File: FoodPin/FoodPin/Controller/RestaurantTableViewController.swift [hunk 5]
```diff
@@ -59,17 +59,13 @@ class RestaurantTableViewController: UITableViewController {
...
         var snapshot = NSDiffableDataSourceSnapshot<Section, Restaurant>()
         snapshot.appendSections([.all])
         snapshot.appendItems(restaurants, toSection: .all)
         dataSource.apply(snapshot, animatingDifferences: false)
-
-        tableView.dataSource = dataSource
         tableView.separatorStyle = .none
         tableView.cellLayoutMarginsFollowReadableWidth = true
-
-        navigationController?.navigationBar.prefersLargeTitles = true
```
Comment (paulpop10) [inline FoodPin/FoodPin/Controller/RestaurantTableViewController.swift:72]: I guess you deleted this line by mistake

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+//  FoodPin
+//
+//  Created by Andreea Sburlea on 28.04.2022.
+//
+
+import UIKit
+
+class RestaurantDetailTwoColumnCell: UITableViewCell {
+
+    @IBOutlet private var column1TitleLabel: UILabel! {
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift]: Any UI elements and objects in general should have intuitive names. You can use something like "addressTitleLabel"

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        didSet {
+            column1TitleLabel.text = column1TitleLabel.text?.uppercased()
+            column1TitleLabel.numberOfLines = 0
+        }
+    }
+    func configureColumn1TitleLabel(text: String) {
+        column1TitleLabel.text = text
+    }
+
+    @IBOutlet private var column1TextLabel: UILabel! {
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift]: Use "addressLabel" or something. This applies to the outlets from below as well

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+class RestaurantDetailTwoColumnCell: UITableViewCell {
+
+    @IBOutlet private var column1TitleLabel: UILabel! {
+        didSet {
+            column1TitleLabel.text = column1TitleLabel.text?.uppercased()
+            column1TitleLabel.numberOfLines = 0
+        }
+    }
+    func configureColumn1TitleLabel(text: String) {
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift]: "configureAddressTitleLabel"

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        column1TitleLabel.text = text
+    }
+
+    @IBOutlet private var column1TextLabel: UILabel! {
+        didSet {
+            column1TextLabel.numberOfLines = 0
+        }
+    }
+    func configureColumn1TextLabel(text: String) {
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift]: same here and for the methods below

File: FoodPin/FoodPin/View/RestaurantDetailTextCell.swift [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+class RestaurantDetailTextCell: UITableViewCell {
+
+    @IBOutlet private var descriptionLabel: UILabel! {
+        didSet {
+            descriptionLabel.numberOfLines = 0
+        }
+    }
+
+    func congigureTextCell(text: String) {
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTextCell.swift]: .

File: FoodPin/FoodPin/Model/Restaurant.swift [hunk 11]
```diff
@@ -56,10 +77,12 @@
          }
      }
 
-    public init(name: String, type: String, location: String, image: String, isFavorite: Bool) {
+     public init(name: String, type: String, location: String, phone: String, description: String, image: String, isFavorite: Bool) {
```
Comment (paulpop10) [inline FoodPin/FoodPin/Model/Restaurant.swift]: unwanted space added

File: FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    @IBOutlet private var heartButton: UIButton!
+    @IBOutlet private var nameLabel: UILabel! {
+        didSet {
+            nameLabel.numberOfLines = 0
+
+            if let customFont = UIFont(name: "Nunito-Bold", size: 40.0) {
+                nameLabel.font = UIFontMetrics(forTextStyle: .title1).scaledFont(for: customFont)
+            }
+        }
+    }
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift:21]: empty space needed

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+import UIKit
+
+class RestaurantDetailTwoColumnCell: UITableViewCell {
+
+    @IBOutlet private var column1TitleLabel: UILabel! {
+        didSet {
+            column1TitleLabel.text = column1TitleLabel.text?.uppercased()
+            column1TitleLabel.numberOfLines = 0
+        }
+    }
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift:17]: empty space needed

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    func configureColumn1TitleLabel(text: String) {
+        column1TitleLabel.text = text
+    }
+
+    @IBOutlet private var column1TextLabel: UILabel! {
+        didSet {
+            column1TextLabel.numberOfLines = 0
+        }
+    }
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift:27]: .

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    func configureColumn1TextLabel(text: String) {
+        column1TextLabel.text = text
+    }
+
+    @IBOutlet private var column2TitleLabel: UILabel! {
+        didSet {
+            column2TitleLabel.text = column2TitleLabel.text?.uppercased()
+            column2TitleLabel.numberOfLines = 0
+        }
+    }
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift:38]: .

File: FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+    func configureColumn2TitleLabel(text: String) {
+        column2TitleLabel.text = text
+    }
+
+    @IBOutlet private var column2TextLabel: UILabel! {
+        didSet {
+            column2TextLabel.numberOfLines = 0
+        }
+    }
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTwoColumnCell.swift:48]: .

File: FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift [hunk 17]
```diff
@@ -8,21 +8,21 @@
...
+        tableView.dataSource = self
+        tableView.separatorStyle = .none
 
-        navigationItem.largeTitleDisplayMode = .never
-        restaurantImageView.image = UIImage(named: restaurant.image)
-        restaurantName.text = restaurant.name
-        restaurantType.text = restaurant.type
-        restaurantLocation.text = restaurant.location
+        // Configure header view
+        headerView.configureHeader(image: restaurant.image, isFavorite: restaurant.isFavorite, name: restaurant.name, type: restaurant.type)
```
Comment (ralumarusca) [inline FoodPin/FoodPin/Controller/RestaurantDetailViewController.swift]: Here you can pass the entire restaurant object to the configureHeader method

File: FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    @IBOutlet var typeLabel: UILabel! {
```
Comment (ralumarusca) [inline FoodPin/FoodPin/View/RestaurantDetailHeaderView.swift]: private IBOutlet :D

**Review by ralumarusca — CHANGES_REQUESTED** [review]

File: FoodPin/FoodPin/Model/Restaurant.swift (hunk 11 above)
Comment (paulpop10) [inline FoodPin/FoodPin/Model/Restaurant.swift]: This issue doesn't seem to be resolved

File: FoodPin/FoodPin/View/RestaurantDetailTextCell.swift [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+
+class RestaurantDetailTextCell: UITableViewCell {
+
+    @IBOutlet private var descriptionLabel: UILabel! {
+        didSet {
+            descriptionLabel.numberOfLines = 0
+        }
+    }
+
+    func congigureDescriptionCell(text: String) {
```
Comment (paulpop10) [inline FoodPin/FoodPin/View/RestaurantDetailTextCell.swift]: typo "configure"

**Review by paulpop10 — APPROVED** [review]

**Review by ralumarusca — APPROVED** [review]

**Pull request merged**